File: 2429-minimize-xor/2429-minimize-xor.py
class Solution:
    def minimizeXor(self, num1: int, num2: int) -> int:
        # Initialize result to num1. We'll modify result.
        res = num1

        target_set_bits_count = bin(num2).count("1")
        set_bits_count = bin(res).count("1")

        # Start with the least significant bit (bit 0).
        curr_bit = 0

        # Add bits to result if it has fewer set bits than the target.
        while set_bits_count < target_set_bits_count:
            # If the current bit in result is not set (0), set it to 1.
            if not self._is_set(res, curr_bit):
                res = self._set_bit(res, curr_bit)
                set_bits_count += 1
            
            # Move to the next bit.
            curr_bit += 1
        
        # Remove bits from result if it has more set bits than the target.
        while set_bits_count > target_set_bits_count:
            # If the current bit in result is set (1), unset it (make it 0).
            if self._is_set(res, curr_bit):
                res = self._unset_bit(res, curr_bit)
                set_bits_count -= 1
            
            # Move to the next bit.
            curr_bit += 1

        return res

    # Helper function to check if the given bit position in result is set (1).
    def _is_set(self, x: int, bit: int) -> bool:
        return (x & (1 << bit)) != 0
    
    # Helper function to set the given bit position in result to 1.
    def _set_bit(self, x: int, bit: int) -> int:
        return x | (1 << bit)
    
    # Helper function to unset the given bit position in x (X set it to 0).
    def _unset_bit(self, x: int, bit: int) -> int:
        return x & ~(1 << bit)

File: 2429-minimize-xor/2429-minimize-xor.cpp
class Solution {
public:
    int minimizeXor(int num1, int num2) {
        // Initialize result to num1. We'll modify result.
        int res = num1;
        
        int targetSetBitsCount = __builtin_popcount(num2), setBitsCount = __builtin_popcount(num1);
        
        // Start with the least significant bit (bit 0).
        int currentBit = 0;

        // Add bits to result if it has fewer set bits than the target.
        while (setBitsCount < targetSetBitsCount){
            // If the current bit in result is not set (0), set it to 1.
            if (!isSet(res, currentBit)){
                setBit(res, currentBit);
                ++setBitsCount;
            }

            // Move to the next bit.
            ++currentBit;
        }

        // Remove bits from result if it has more set bits than the target.
        while (setBitsCount > targetSetBitsCount){
            // If the current bit in result is set (1), unset it (make it 0).
            if (isSet(res, currentBit)){
                unsetBit(res, currentBit);
                --setBitsCount;
            }

            // Move to the next bit.
            ++currentBit;
        }
        
        return res;
    }

private:
    // Helper function to check if the given bit position in x is set (1).
    bool isSet(int x, int bit){
        return (x & (1 << bit)) != 0;
    }

    // Helper function to set the given bit position in x to 1.
    void setBit(int &x, int bit){
        x |= (1 << bit);
    }

    // Helper function to unset the given bit position in x (set it to 0).
    void unsetBit(int &x, int bit){
        x &= ~(1 << bit);
    }
};